tools/codegen/src/lib.rs

```diff
@@ -0,0 +1,383 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use std::{
+    cmp::{self, Reverse},
+    collections::BTreeMap,
+    env, fmt,
+    path::{Path, PathBuf},
+    slice,
+    str::FromStr,
+};
+
+use anyhow::Result;
+use serde::{
+    de::{self, Deserialize, Deserializer},
+    ser::{Serialize, Serializer},
+};
+use serde_derive::{Deserialize, Serialize};
+
+#[must_use]
+pub fn workspace_root() -> PathBuf {
+    let mut dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+    dir.pop(); // codegen
+    dir.pop(); // tools
+    dir
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Version {
+    pub major: Option<u64>,
+    pub minor: Option<u64>,
+    pub patch: Option<u64>,
+    pub pre: semver::Prerelease,
+    pub build: semver::BuildMetadata,
+}
```

Comment:
fwiw, `git diff --color-moved` is a good way to check changes to the items which were moved to lib.rs.

Or if you like, I can create a separate PR to move them to lib.rs , but it still involves a lot of changes because `pub` needed to be added to the struct members